src/box/lua/space.cc

```diff
@@ -772,12 +772,16 @@ box_lua_space_delete(struct lua_State *L, uint32_t id)
 static int
 box_lua_space_new_or_delete(struct trigger *trigger, void *event)
 {
-	struct lua_State *L = (struct lua_State *) trigger->data;
+	(void)trigger;
+	struct lua_State *L = tarantool_L;
 	struct space *space = (struct space *) event;
 
 	if (space_by_id(space->def->id) != NULL) {
 		box_lua_space_new(L, space);
 	} else {
+		/* tarantool_L is freed on Tarantool shutdown. */
+		if (L == NULL)
+			return 0;
```

Comment:
We should delete the trigger in `box_lua_space_free` instead.

Reply:
Fixed.

<details>
<summary>diff</summary>

```diff
diff --git a/src/box/lua/init.c b/src/box/lua/init.c
index 6bdc27a186..5d607b4071 100644
--- a/src/box/lua/init.c
+++ b/src/box/lua/init.c
@@ -995,4 +995,5 @@ void
 box_lua_free(void)
 {
 	box_lua_iproto_free();
+	box_lua_space_free();
 }
diff --git a/src/box/lua/space.cc b/src/box/lua/space.cc
index ce9830f206..79276fadad 100644
--- a/src/box/lua/space.cc
+++ b/src/box/lua/space.cc
@@ -772,16 +772,12 @@ box_lua_space_delete(struct lua_State *L, uint32_t id)
 static int
 box_lua_space_new_or_delete(struct trigger *trigger, void *event)
 {
-	(void)trigger;
-	struct lua_State *L = tarantool_L;
+	struct lua_State *L = (struct lua_State *) trigger->data;
 	struct space *space = (struct space *) event;
 
 	if (space_by_id(space->def->id) != NULL) {
 		box_lua_space_new(L, space);
 	} else {
-		/* tarantool_L is freed on Tarantool shutdown. */
-		if (L == NULL)
-			return 0;
 		box_lua_space_delete(L, space->def->id);
 	}
 	return 0;
@@ -921,7 +917,7 @@ void
 box_lua_space_init(struct lua_State *L)
 {
 	/* Register the trigger that will push space data to Lua. */
-	assert(tarantool_L == L);
+	on_alter_space_in_lua.data = L;
 	trigger_add(&on_alter_space, &on_alter_space_in_lua);
 
 	lua_getfield(L, LUA_GLOBALSINDEX, "box");
@@ -1029,3 +1025,9 @@ box_lua_space_init(struct lua_State *L)
 	luaL_setfuncs(L, space_internal_lib, 0);
 	lua_pop(L, 1);
 }
+
+void
+box_lua_space_free(void)
+{
+	trigger_clear(&on_alter_space_in_lua);
+}
diff --git a/src/box/lua/space.h b/src/box/lua/space.h
index 1155e8fb39..7421332de0 100644
--- a/src/box/lua/space.h
+++ b/src/box/lua/space.h
@@ -41,6 +41,10 @@ struct lua_State;
 void
 box_lua_space_init(struct lua_State *L);
 
+/** Free module resources. */
+void
+box_lua_space_free(void);
+
 #if defined(__cplusplus)
 } /* extern "C" */
 #endif /* defined(__cplusplus) */

```

</details>